Add Mtime.absolute for log correlation use cases

dsheets opened the pull request:
This patch adds `Mtime.absolute` which allows processes (or scripts) to share a global reference moment. This, in turn, makes the library usable for log or other timeline correlation between non-cooperating actors.

Please, let me know if there are additional tasks to be done or if you'd like this feature designed differently.

dsheets committed:
Add Mtime.absolute for log correlation use cases

dbuenzli commented:
I think I pondered quite a while about providing this when I designed the module in the first place and IIRC I discarded the idea because: 
1. You can't really store the numbers since they would become meaningless after a reboot.
2. The APIs don't say anything about the absolute magnitude of the monotonic clock (in particular they don't say that it starts at 0 at boot time). As you can see in the code I diff the time data structures before cramming their bits in a single `uint64`. So the current PR has risks of overflow, which means that we can't really rely on it and I'm not willing to provide something you can't rely on.

Regardless of 2. which we could e.g. solve by providing another data type. I'd be interested in having a compelling use case for this especially w.r.t. 1. The only way I see this working is if a process coordinates the logging processes in which case the current support is enough.

dbuenzli commented:
Note that one way to side step point 2. would be to put these problems in the hands of the client... which we could do by simply exposing a `Mtime.absolute : span` with the value of the C static variable `start` of the current implementation. Together with `Mtime.elapsed` this would allow the client to recover your `Mtime.absolute` and let it handle overflow problems.

dsheets commented:
1. You _can_ store the numbers because they are meaningful relative to each other in a single session.
2. A `uint64_t` has range for more than 272 years at nanosecond precision. If the system APIs risk overflow in a `uint64_t` to store internal time, they are in error and a risk to all client code using them. Additionally, the systems in question _do_, in practice, start at zero or near-zero at some initialization moment.

Process coordination is a non-starter when the processes have no way to share clocks save the system's shared clock functionality. Additionally, use of the shared timepiece allows decoupled processes to correlate their logs inside of a single system 'session' (e.g. boot, page load, etc).

I do think that `start` could be usefully exposed but I really don't care about it in this case and involving it in the process of using the system monotonic clock seems unnecessarily complicated. I really just want access to the system's highest precision monotonic time source.

dbuenzli commented:
> You can store the numbers because they are meaningful relative to each other in a single session.

But then you also need a way to detect when a new session begins to invalidate the old stamps. So I wonder if mtime is the right place to solve this problem.

> If the system APIs risk overflow in a uint64_t to store internal time

The system APIs (at least POSIX) do not take this risk, they use a uint64 for seconds and then a uint64 for the number of nanoseconds in the last second.

> Process coordination is a non-starter when the processes have no way to share clocks save the system's shared clock functionality.

That's wrong. Your processes can all talk to the same coordinating process (e.g. `syslogd`) or they can all be forked from the same parent process. In both these cases `Mtime.elapsed` is sufficient.

dsheets commented:
> But then you also need a way to detect when a new session begins to invalidate the old stamps. So I wonder if mtime is the right place to solve this problem.

The old stamps don't become invalid when a new session begins. They are simply not comparable to times from the new session's epoch. `mtime` is not the right place to solve the session problem which is why I desire a lower level API than `elapsed` which forces a solution to the session problem on the user (you can only get times relative to the `elapsed` epoch and the system-level epoch information is destroyed).

> The system APIs (at least POSIX) do not take this risk, they use a uint64 for seconds and then a uint64 for the number of nanoseconds in the last second.

POSIX uses `timespec` because it always uses `timespec`. The current implementation of `elapsed` does not account for roll-over. Even if the time returned from the monotonic clock is based on the UNIX epoch, we will still have a good 226 years give or take before the `uint64_t` overflows.

> That's wrong. Your processes can all talk to the same coordinating process (e.g. syslogd) or they can all be forked from the same parent process. In both these cases Mtime.elapsed is sufficient.

It's not _wrong_, it's a design constraint (_when_ is the operative word). The processes need to be able to write data to disk which contains log events which can later be correlated between different process logs. They are separate executables and cannot share the `mtime` epoch. They do not and should not have to talk to a logging daemon.

dsheets commented:
I've updated the commit to include `static` on `scale`. I don't know what happened to your line comment.

Here are some ideas about conversion and overflow:
- We could delay conversion into nanoseconds until `span` is destructed to allow unconverted arithmetic.
- We could offer the user access to the system's time units.
- We could tell the user how big the numerator of the system's nanosecond conversion factor is so they know how many bits less than 64 they have for the system's ticks.
- We could provide a way to query for "safe" durations or offer the longest "safe" duration value.

avsm commented:
Just checking in on this PR @dbuenzli, any opinions on progressing it?

dbuenzli commented:
I'm okay to add something along these lines but I'm not convinced by the API. First the `span` value returned by the proposed function does not satisfy what is documented under the `span` type. Second it completely invalidates the module preamble documentation which is quite suspicious.

This actually makes me wonder wether we really want this as proposed here. I think we should rather have an `Abs` sub-module with something like: 

``` ocaml
module Abs : sig
  type t 

  val now : unit -> t 

  val equal : t -> t -> bool 
  val compare : t -> t -> int
  val is_earlier : t -> than:t -> bool
  val is_later : t -> than:t -> bool

  val diff : t -> t -> span
  val add_span : t -> span -> t option
  val sub_span : t -> span -> t option

  val to_raw : t -> int64
  val of_raw : int64 -> t 
  val pp : Format.formatter -> t -> unit 
end
```

dsheets committed:
Introduce new System submodule and type for direct monotonic clock values

dsheets committed:
Fix comparison of concrete int64 to behave as uint64

dsheets committed:
Detect underflow/overflow in System.sub_span/System.add_span

dsheets committed:
Fix System.span to obey uint64 semantics

dsheets committed:
Fix underflow/overflow in jsoo System.sub_span/System.add_span

dsheets committed:
Add comparison test

dsheets committed:
Replace Mtime.compare with C function

dsheets committed:
Add test_jsoo.* to .gitignore

dsheets commented:
I've updated this PR to follow your suggestion and add a new submodule, `System`, which provides direct access to the monotonic clock. I called your `Abs.diff` function `System.span` and your `Abs.{to,of}_raw` `System.{to,of}_ns_uint64` to align with the functions of the same name under `Mtime`.

I also realized that `compare` introduced in #18 was buggy in cases where timestamps are very large and the (un-)signedness of the `int64` value makes a difference. I implemented an unsigned comparison in OCaml in c2b2180 and an unsigned comparison in C in 2778bb5 (cc @yallop to take another look and ensure that the C is copasetic). I added a test for this behavior as well.

I'm happy to squash or reword or change anything about this PR that you'd like different. This is the last patch that we're carrying on `mtime` (after #18 and #19) and we'd really like to depend on a mainline release of the library. Thanks!

yallop commented:
> cc @yallop to take another look and ensure that the C is copasetic

The interaction with the runtime (`CAMLparam*` etc.) looks good to me.  I haven't looked carefully at the comparison semantics.

dsheets commented:
Any chance this can be reviewed soon?

dbuenzli commented:
Sorry I know I have been lagging too much on this, will try to resolve this in the upcoming weeks.

dbuenzli commented:
I merged the patches on master though as we discussed I'm going to reorganize this a bit.
